In [ ]:
!python -m pip install statsmodels

## Глоссарий

- Мода (mode): Самое часто встречающееся значение в выборке, т.е. значение, которое появляется наиболее часто. Например, если в выборке {3, 5, 5, 6, 6, 6, 9}, то модой будет 6, так как она встречается чаще всего. 

- Медиана (median): Центральное значение выборки, которое разделяет выборку на две равные по размеру части. Для этого выборка сортируется по возрастанию (или убыванию), и медианой будет серединное значение. Например, для выборки {3, 6, 7, 8, 8, 10, 14}, медиана будет равна 8.

- Среднее (mean): Среднее значение в выборке получается путем суммирования всех значений и деления на их количество. Например, для выборки {2, 4, 4, 4, 5, 5, 7, 9} среднее значение будет равно (2+4+4+4+5+5+7+9)/8 = 5.

- Дисперсия (variance): Дисперсия показывает, насколько значения в выборке отклоняются от их среднего значения. Большая дисперсия означает, что значения сильно разбросаны вокруг среднего, а маленькая дисперсия говорит о более плотном скоплении значений вокруг среднего.

- Выборка (sample): Набор данных, полученных из генеральной совокупности, который используется для проведения статистических анализов.

- Стандартное отклонение (standard deviation): Мера разброса данных в выборке относительно их среднего значения. Оно показывает, насколько точные значения "сгруппированы" вокруг среднего значения.

- Стандартная ошибка (standard error): Мера неопределенности или ошибки в оценке параметра по выборке. Чем меньше стандартная ошибка, тем более точно выборочное среднее оценивает среднее значение генеральной совокупности.

- Доверительный интервал (confidence interval): Диапазон значений, в который с заданной вероятностью попадает параметр генеральной совокупности.

- Альфа (α): Уровень значимости, используемый в статистических тестах, который определяет вероятность ошибки первого рода.

- Мощность (power): Вероятность обнаружения статистически значимых различий, если они существуют.

- Z-критерий (Z-test): Статистический тест, используемый для проверки гипотез относительно средних значений, когда известно стандартное отклонение генеральной совокупности или объем выборки большой.

- T-критерий (t-test): Статистический тест, используемый для проверки гипотез относительно средних значений, когда стандартное отклонение генеральной совокупности неизвестно и должно быть оценено по выборке.


In [ ]:
import numpy as np
from scipy import stats

In [2]:
# Ваш набор данных
data = [5, 5, 5, 18, 10, 8, 9, 10, 19, 21]

# Вычисляем выборочное среднее
mean = np.mean(data)

# Выборочная дисперсия
variance = np.var(data, ddof=1)

# Стандартное отклонение
std_dev = np.std(data, ddof=1)

# Медиана
median = np.median(data)

# Мода
mode = stats.mode(data)

print("Выборочное среднее:", mean)
print("Выборочная дисперсия:", variance)
print("Стандартное отклонение:", std_dev)
print("Медиана:", median)
print("Мода:", mode)

Выборочное среднее: 11.0
Выборочная дисперсия: 37.333333333333336
Стандартное отклонение: 6.110100926607787
Медиана: 9.5
Мода: ModeResult(mode=array([5]), count=array([3]))


C:\Users\admin\AppData\Local\Temp/ipykernel_9656/3537758302.py:17: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode = stats.mode(data)


In [2]:
import scipy.stats as stats
import math

# Заданные параметры
conversion_rate = 0.11  # Конверсия
sample_size = 1500  # Размер выборки
confidence_level = 0.95  # Уровень доверия

# Расчет стандартной ошибки
standard_error = math.sqrt((conversion_rate * (1 - conversion_rate)) / sample_size)

# Расчет доверительного интервала
margin_of_error = stats.norm.ppf((1 + confidence_level) / 2) * standard_error
lower_bound = conversion_rate - margin_of_error
upper_bound = conversion_rate + margin_of_error

# Вывод результатов
print(f"95% доверительный интервал для конверсии: [{lower_bound:.4f}, {upper_bound:.4f}]")

95% доверительный интервал для конверсии: [0.0942, 0.1258]


## ЗАДАЧА:

Дизайнер предложил добавить рекомендательную систему на этапе формирования корзины. Нынешний средний чек равен 2167 рублей, а стандартная ошибка (SD) равна 69. 
Мы предполагаем, что в новой версии средний чек вырастет 2180. Определите, сколько юзеров нам нужно отправить на каждую версию с уровнем доверия 95%

In [46]:
import math
import statsmodels.stats.power
from statsmodels.stats.power import TTestIndPower

mean_c = 2167  # Среднее значение для контрольной группы
mean_t = 2180  # Среднее значение для тестовой группы
sd = 69  # Стандартное отклонение
alpha = 0.05  # Уровень значимости
power = 0.95  # Мощность (вероятность обнаружения эффекта)

effect_size = mean_c/ mean_t/ sd  # Вычисление размера эффекта
print(f"Эффект= {effect_size:.2f}")  # Вывод размера эффекта

ratio=1  # Предполагаемое отношение размеров выборок

peremennaja = TTestIndPower()  # Инициализация анализа мощности для независимых выборок
razmer_viborka = peremennaja.solve_power(effect_size = effect_size, alpha=alpha, power = power, ratio=ratio)  # Определение необходимого размера выборки
print(f"Размер выборки {razmer_viborka:.2f}")  # Вывод полученного размера выборки


Эффект= 0.01
Размер выборки 125225.65


# ЗАДАЧА
Вы решили сравниваем метрику деньги на юзера в двух группах. Размер выборки - 1000 элементов в каждой группе. 
Для проверки нормальности распределения на выборке в 1000 наблюдений применили , критерий Шапиро-Уилка и получили p-value, равный 0.00002, alpha = 5% (0.05) то какой бы вывод мы могли сделать в данном случае?
В этом случае какой статистический критерий для проверки первоначальной гипотезы тут лучше всего подойдёт и почему ?


In [19]:
from scipy.stats import shapiro, mannwhitneyu
import numpy as np

# Заданные вводные данные
alpha = 0.05
group_size = 1000

# Генерация данных для двух групп (пусть это будут примеры случайных данных, так как задача не содержит таких данных)
group1 = np.random.normal(loc=100, scale=20, size=group_size)
group2 = np.random.normal(loc=105, scale=20, size=group_size)

# Проверка нормальности распределения
_, p_value1 = shapiro(group1)
_, p_value2 = shapiro(group2)

# Вывод результатов теста на нормальность
print(f"Группа 1: p-value критерия Шапиро-Уилка = {p_value1:.5f}")
print(f"Группа 2: p-value критерия Шапиро-Уилка = {p_value2:.5f}")

# Проведение теста Манна-Уитни для сравнения метрик между группами
_, p_value_mannwhitney = mannwhitneyu(group1, group2)

# Вывод результатов теста Манна-Уитни
print("\nТест Манна-Уитни:")
print(f"p-value = {p_value_mannwhitney:.5f}")

# Проверка статистической значимости
if p_value_mannwhitney < alpha:
    print("\nОтклоняем нулевую гипотезу. Существует статистически значимая разница между группами.")
else:
    print("\nНе отклоняем нулевую гипотезу. Разница между группами статистически не значима.")

Группа 1: p-value критерия Шапиро-Уилка = 0.60239
Группа 2: p-value критерия Шапиро-Уилка = 0.75206

Тест Манна-Уитни:
p-value = 0.00000

Отклоняем нулевую гипотезу. Существует статистически значимая разница между группами.


## ЗАДАЧА:

Сравниваем метрику конверсия в покупку. Размер выборки - 10000 элементов в каждой группе . Какой(ие) статистический(ие) критерий(ие) тут лучше всего подойдёт(ут) и почему ?

Размер выборки большой, поэтому используем Z критерий или Хи квадрат





In [51]:
import numpy as np
from statsmodels.stats.proportion import proportions_ztest
from scipy.stats import chisquare  # импортируем функцию для Хи-квадрат теста из scipy.stats

# Заданные вводные данные
group_size = 10000

# Предположим, у вас есть количество покупок и общее количество посещений для двух групп
successes_group1 = 300  # количество покупок в группе 1
nobs_group1 = 10000  # общее количество посещений в группе 1

successes_group2 = 350  # количество покупок в группе 2
nobs_group2 = 10000  # общее количество посещений в группе 2

# Выполнение z-теста для разности долей
stat, p_value = proportions_ztest([successes_group1, successes_group2], [nobs_group1, nobs_group2])

# Вывод результатов z-теста
print(f"Z-статистика: {stat:.4f}")
print(f"p-value: {p_value:.4f}")

# Проверка статистической значимости для z-теста
alpha = 0.05
if p_value < alpha:
    print("Отклоняем нулевую гипотезу. Существует статистически значимая разница в конверсии между группами.")
else:
    print("Не отклоняем нулевую гипотезу. Разница в конверсии между группами статистически не значима.")

# Хи-квадрат тест
observed = np.array([[successes_group1, nobs_group1 - successes_group1], [successes_group2, nobs_group2 - successes_group2]])
chi_stat, chi_p_value = chisquare(observed)

print(f"Хи-квадрат тест: статистика={chi_stat}, p-value={chi_p_value}")


Z-статистика: -1.9938
p-value: 0.0462
Отклоняем нулевую гипотезу. Существует статистически значимая разница в конверсии между группами.
Хи-квадрат тест: статистика=[3.84615385 0.12919897], p-value=[0.0498602  0.71926403]


## ХИ - квадрат

1. Для статистики:
   - Для первой группы (статистика=3.84615385): это значение показывает, насколько сильно отклоняются наблюдаемые данные от ожидаемых. Чем больше значение этой статистики, тем сильнее отклонение.
   - Для второй группы (статистика=0.12919897): аналогично, это значение отражает силу отклонения данных от ожидаемых, при условии проведения Хи-квадрат теста для второй группы.

2. Для p-value:
   - Для первой группы (p-value=0.0498602): это значение вероятности получить статистику, равную или большую наблюдаемой, при условии верности нулевой гипотезы. Таким образом, небольшое значение p-value указывает на статистическую значимость отклонения.
   - Для второй группы (p-value=0.71926403): это значение вероятности получить статистику, равную или большую наблюдаемой, при условии верности нулевой гипотезы для второй группы.

Исходя из этих результатов, в первой группе можно сделать вывод, что отличие статистически значимо (p-value меньше уровня значимости 0.05), в то время как во второй группе такого отличия обнаружено не было (p-value значительно больше уровня значимости).